sdk/identity/azure-identity/inc/azure/identity/detail/client_credential_core.hpp

```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "azure/identity/dll_import_export.hpp"
+
+#include <azure/core/credentials/credentials.hpp>
+#include <azure/core/url.hpp>
+
+#include <string>
+
+namespace Azure { namespace Identity { namespace _detail {
+  class ClientCredentialCore final {
+    Core::Url m_authorityHost;
+    std::string m_tenantId;
+    bool m_isAdfs;
+
+  public:
+    AZ_IDENTITY_DLLEXPORT static std::string const AadGlobalAuthority;
+
+    explicit ClientCredentialCore(std::string tenantId, std::string const& authorityHost);
+
+    Core::Url GetRequestUrl() const;
+
+    std::string GetScopesString(decltype(Core::Credentials::TokenRequestContext::Scopes)
```

Comment:
Can we  use a regular function signature declaration here rather than `decltype`?

Reply:
We always can put a real type, but I like decltypes - if type gets changed in the `TokenRequestContext`, everything keeps working, no need to update the type here. Decltype allows me to say "Whatever the type `TokenRequestContext::Scopes` is". It can make sense to avoid it in user-facing APIs for simplicity, but for the internal code, quite the opposite.

Reply:
Though every dev has preferences for implementation details and empowered to choose what C++ patterns to follow, I think it is valuable to keep the entire code base consistent for maintenance (similar to, but not necessarily to the same extent as public API surface).

I see the value of `decltype` in some dynamically typed scenarios, but **to me**, it makes the code more complex to read and understand. Therefore, I would try to minimize its use within the SDK.

`TokenRequestContext` is a public type and won't ever change (as that's breaking). I don't think it makes sense to use `decltype` here:
https://github.com/Azure/azure-sdk-for-cpp/blob/4de2423934cc5526bcca9439d5e643052d3201db/sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp#L44

Reply:
I respectfully disagree, decltype() is the best tool for the job, especially when all the function does is passes it to another function. When you read this function, in no place you need to know its actual type, that it is actually a vector. It gets passed further. Using decltype() where it was meant to be does actually mean easier maintenance, not harder. I don't find decltype() hard to read or understand - it is a language feature, and we use it as it was meant to be used, plus we use inside the implementation, and unlike the public API service, users who don't know about this language feature can't expect to read implementation sources and understand everything without further reading.